### PR TITLE
Add extension methods for Reuse[View*]] and some Reuse helper methods.

### PR DESCRIPTION
**Pull request by toddburnside**

### CODE REVIEW

File: js/src/main/scala/crystal/react/reuse/Reuse.scala [hunk 1]
```diff
@@ -33,22 +33,26 @@ import scala.reflect.ClassTag
...
   protected[reuse] implicit val ClassTag: ClassTag[B]
 
   protected[reuse] implicit val reusability: Reusability[B]
 
+  def addReuseBy[R: Reusability](r: R): Reuse[A] = Reuse.by((reuseBy, r))(value)
+  def addReuse[C](r: Reuse[C]): Reuse[A]         = addReuseBy(r.reuseBy)(r.reusability)
+
+  def replaceReuseBy[R: Reusability: ClassTag](r: R): Reuse[A] = Reuse.by(r)(value)
+  def replaceReuse[C](r: Reuse[C]): Reuse[A]                   = replaceReuseBy(r.reuseBy)(r.reusability, r.ClassTag)
+
```
Comment (rpiaggio) [inline js/src/main/scala/crystal/react/reuse/Reuse.scala:51]: I have local changes that may reformulate these. I'll merge and send another PR shortly.

File: js/src/main/scala/crystal/react/reuse/Reuse.scala [hunk 2]
```diff
@@ -33,22 +33,26 @@ import scala.reflect.ClassTag
 trait Reuse[+A] {
   type B
 
-  val get: () => A
+  lazy val value: A = getValue()
 
-  lazy val value: A = get()
+  protected[reuse] val getValue: () => A
```
Comment (rpiaggio) [inline js/src/main/scala/crystal/react/reuse/Reuse.scala:38]: Why was this necessary?

Comment (toddburnside) [reply js/src/main/scala/crystal/react/reuse/Reuse.scala:38]: Not necessary. But we discussed it. The renaming would have been enough. But, it seems like `value` will be the one used, not `getValue`.

File: js/src/main/scala/crystal/react/reuse/package.scala [hunk 3]
```diff
@@ -126,4 +142,122 @@ package object reuse {
...
+
+    def unsafeNarrow[B <: A]: Reuse[ViewF[F, B]] =
+      zoom(_.asInstanceOf[B])(modB => a => modB(a.asInstanceOf[B]))
+
+    def to[F1[_]: Monad](
+      toF1:   F[Unit] => F1[Unit],
+      fromF1: F1[Unit] => F[Unit]
+    ): Reuse[ViewF[F1, A]] = rv.map(_.to[F1](toF1, fromF1))
+
+    def mapValue[B, C](f: ViewF[F, B] => C)(implicit ev: A =:= Option[B]): Option[C] =
```
Comment (rpiaggio) [inline js/src/main/scala/crystal/react/reuse/package.scala:189]: Maybe `f` can be `Reuse[ViewF[F, B]] => C` so we can propagate reusability.

File: js/src/main/scala/crystal/react/reuse/package.scala [hunk 4]
```diff
@@ -126,4 +142,122 @@ package object reuse {
...
+      rv.map(v => get.map(a => f(v.zoom(_ => a)(f => a1 => ev.flip(a1.map(f))))))
+
+    def xxx: Reuse[ViewF[F, A]] = rv
```
Comment (rpiaggio) [inline js/src/main/scala/crystal/react/reuse/package.scala:192]: Was this a test?

File: js/src/main/scala/crystal/react/reuse/Reuse.scala (hunk 2 above)
Comment (toddburnside) [reply js/src/main/scala/crystal/react/reuse/Reuse.scala:38]: The name `get` conflicted with `Reuse,get` wrt the extension methods.

File: js/src/main/scala/crystal/react/reuse/package.scala (hunk 3 above)
Comment (toddburnside) [reply js/src/main/scala/crystal/react/reuse/package.scala:189]: Hmm. I neglected to change the return type on this one (I did in ReuseViewOptF]). The return really is `Reuse[Option[C]]`, but the automatic conversion is changing it. Do you want `Reuse` in the `f` or just the return?

Comment (toddburnside) [reply js/src/main/scala/crystal/react/reuse/package.scala]: If `f` is `Reuse[ViewF[F, B]] => C`, would we combine the reusability of `f` with the current reusability?

File: js/src/main/scala/crystal/react/reuse/Reuse.scala (hunk 2 above)
Comment (rpiaggio) [reply js/src/main/scala/crystal/react/reuse/Reuse.scala:38]: Oh ok sorry. My brain had trouble processing the fact that they switched positions 👍

**Review by rpiaggio — APPROVED** [review]
LGTM, thanks!

File: js/src/main/scala/crystal/react/reuse/package.scala (hunk 3 above)
Comment (rpiaggio) [reply js/src/main/scala/crystal/react/reuse/package.scala]: I think that what's important is that we pass a `Reuse[View[A]]` into the function. We should build that with the same reusability that we have for the `Reuse[View[Option[A]]]`. We can return `Option[C]`.

Use case for this is for example:

``` scala
val view: View[ADTRoot]

<.div(
  view
    .zoom(ADTRoot.prismToOneOfTheSubtypes)
    .mapValue(stillReusableView => Component(stillReusableView))
    .whenDefined
)
```

Comment (toddburnside) [reply js/src/main/scala/crystal/react/reuse/package.scala]: Ah-ha. I was tired and my brain didn't really grok what the method did - I just did something mechanical.
